### ISSUE TITLE
module: Preserve

### ISSUE THREAD

**Issue by fregante** [sindresorhus/tsconfig#29]
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-4.html#support-for-require-calls-in---moduleresolution-bundler-and---module-preserve

I think this is the right config to finally get TS out of the way

```js
{
    "compilerOptions": {
        "module": "preserve",
        // ^ also implies:
        // "moduleResolution": "bundler",
        // "esModuleInterop": true,
        // "resolveJsonModule": true,
        // ...
    }
}
```

More info in https://github.com/microsoft/TypeScript/pull/56785

**Comment by sindresorhus** [comment]
What's the benefit for ESM-only projects? Not sure I see the value in it.

**Comment by fregante** [comment]
I think the benefit is to stop fighting TypeScript with imports: just let the user use what they write. Specific rules/preferences can be enforced via eslint. 

Or at least that’s the promise

**Comment by sindresorhus** [comment]
There is really no fighting if you are pure ESM. And I don't really see the point in allowing `require` here and then use a linter to disallow it, when it can just be disallowed here instead.

**Comment by fregante** [comment]
> There is really no fighting if you are pure ESM

I reached this point because this config blocks import attributes AKA “pure ESM”, so the fight still exists. Then node wants .ts imports and TS wants .js imports. The fight continues.

**Comment by fregante** [comment]
This is what I see with `@sindresorhus/tsconfig` 6, without changes: ESM is now allowed, `require` is:

<img width="719" alt="Screenshot 13" src="https://github.com/user-attachments/assets/355cc2df-9530-43ce-abd4-e60992a0903e">

**Comment by sindresorhus** [comment]
> If preserve keeps its promise to just allow modules as written, I think it’s advantageous.

I suggest trying it out in practice first on some popular projects to see its effects. I'm already lukewarm to it, but I'm definitely not adding it here until it's proven to work.

**Comment by sindresorhus** [comment]
> This is what I see with @sindresorhus/tsconfig 6, without changes: ESM is not allowed, require is:

That's just the fact of using TS in general. Every JS feature has to be supported by TS before you can use it there. I thought TS already supported import attributes: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-3.html

**Comment by fregante** [comment]
> I thought TS already supported import attributes

They do, but you have to use a specific config (and this config particularly blocks `resolveJsonModule` because Node <18.20 doesn't support it)

**Comment by fregante** [comment]
Makes sense, however this config also includes the DOM lib, which is not available on Node.

How about splitting the config into node (default) and "browser" (`@sindresorhus/tsconfig/browser`)

**Comment by silverwind** [comment]
> TS will load all the node_modules/@types/* declaration files when types field is removed

Also to note, this likely won't be true anymore in typescript 6 as per https://github.com/microsoft/TypeScript/issues/54500 where types will default to `[]` and they plan to remove the auto-discovery.

**Comment by zanminkian** [comment]
> This approach won't work correctly for projects that have both web and node files. I think multiple tsconfig files would be needed in such cases to restrict the type loading to the relevant files.

Projects that have both web and node files should install them together:  `npm i @types/node @types/web -D`.
